# ⚙️ ***Analysis of Factors Affecting Student Academic Performance***



## 🧹 **Step 2: Data Pre-processing**

### 🔸 ****Goal:** Clean and prepare the data for accurate analysis.**



Based on the issues identified in Step 1, we will perform the following data cleaning tasks:



### 🔸 ****Tasks to Complete:****



Perform all necessary cleaning steps such as:



- Handling missing values



- Removing duplicates



- Correcting data types



- Text Cleaning



- Creating derived columns



- Feature transformation

In [2]:
import pandas as pd
import numpy as np
#Display settings
pd.set_option('display.max_columns', None)              #show all columns
pd.set_option('display.float_format','{:.2f}'.format)    #format all decimals to 2 places

print("Libraries imported successfully!")

Libraries imported successfully!


In [28]:
#Load the dataset
df = pd.read_csv(r"C:\Users\anton\Downloads\archive\Student_Performance_Cleaned.csv", encoding='latin-1')

# Display dataset dimensions
print("Dataset loaded successfully!")
print("="*70)
print(f"Dataset Shape: {df.shape[0]:,} rows × {df.shape[1]} columns")

Dataset loaded successfully!
Dataset Shape: 649 rows × 33 columns


## 🔸 **2.1 Handling Missing Values**

### 🔸 **# Check for missing values:**

In [29]:
#used to identify missing (null) values in each column of the dataset.
df.isnull().sum()

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
dtype: int64

The output shows 0 for every column, including variables like school, sex, age, Medu, Fedu, studytime, absences, G1,G2 and G3.

### 🔸 **# Handle missing values (based on observation):**

In [12]:
#deletes rows (by default) that have at least one missing value
before_rows = df.shape[0]
df.dropna(inplace=True)
after_rows = df.shape[0]

print(f"Rows before dropna: {before_rows}")
print(f"Rows after dropna: {after_rows}")

Rows before dropna: 649
Rows after dropna: 649


since the dataset had no difference before or after, no rows were dropped and the data remained unchanged.

## 🔸 **2.2 Removing Duplicates**

In [25]:
# Check number of duplicate rows
duplicate_count = df.duplicated().sum()
print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 0


In [27]:
# Remove duplicate rows if any
df.drop_duplicates(inplace=True)
# Recheck duplicates
print("Duplicates after removal:", df.duplicated().sum())

Duplicates after removal: 0


## 🔸 **2.3 Correcting Data Types**

### 🔸 **# Checking for Correct Date Format**

In [ ]:
#df['Order Date'] = pd.to_datetime(df['Order Date'])
#Since there are no existing columns with dates this step is not needed for this dataset

### 🔸 **# Converting Data Types**

In [23]:
# Converting age column to integer
df['age'] = df['age'].astype(int)

# Verifying data type conversion
print(df['age'].dtype)


int64


As you can see Changes were made Correctly

## 🔸 **2.4 Text Cleaning**

### 🔸 **# Clean Text Columns**

In [24]:
df['school'] = df['school'].str.lower().str.strip()
print("Unique values in school column:")
print(df['school'].unique())


Unique values in school column:
['gp' 'ms']


## 🔸 **2.5 Creating Derived Columns**

### 🔸 **# Average Scores**

A new feature, avg_score, was created by calculating the mean of students’ first, second, and final period grades (G1, G2, and G3). This feature represents the overall academic performance of each student.

In [17]:
# Creating average score feature
df['avg_score'] = df[['G1','G2','G3']].mean(axis=1)

# Verifying feature creation
df[['G1','G2','G3','avg_score']].head()

G1  G2  G3  avg_score
0   0  11  11       7.33
1   9  11  11      10.33
2  12  13  12      12.33
3  14  14  14      14.00
4  11  13  13      12.33

### 🔸 **# Total Score**

A new feature, total_score, was created by summing students’ first, second, and final period grades (G1, G2, and G3). This feature represents the cumulative academic performance of each student.

In [18]:
# Creating total score feature
df['total_score'] = df['G1'] + df['G2'] + df['G3']

# Verifying feature creation
df[['G1', 'G2', 'G3', 'total_score']].head()


G1  G2  G3  total_score
0   0  11  11           22
1   9  11  11           31
2  12  13  12           37
3  14  14  14           42
4  11  13  13           37

### 🔸 **# Result**

A new categorical feature, result, was created based on the final grade (G3). Students scoring 10 or above were labeled as “Pass”, while those scoring below 10 were labeled as “Fail”. This column can be used for classification analysis.

In [19]:
# Creating pass/fail result column based on final grade
df['result'] = df['G3'].apply(lambda x: 'Pass' if x >= 10 else 'Fail')

# Verifying result classification
df[['G3', 'result']].head(10)


G3 result
0  11   Pass
1  11   Pass
2  12   Pass
3  14   Pass
4  13   Pass
5  13   Pass
6  13   Pass
7  13   Pass
8  17   Pass
9  13   Pass

## 🔸 **2.6 Feature Transformation**

### 🔸 **# Normalization**

Min–Max scaling was applied to the age and absences columns to normalize their values between 0 and 1. This ensures that these features are on a comparable scale and prevents features with larger ranges from dominating the analysis.

In [20]:
# Applying MinMax scaling to age and absences
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
df[['age','absences']] = scaler.fit_transform(df[['age','absences']])

# Verifying scaling
df[['age','absences']].describe()

age  absences
count 649.00    649.00
mean    0.00      0.11
std     0.04      0.15
min     0.00      0.00
25%     0.00      0.00
50%     0.00      0.06
75%     0.00      0.19
max     1.00      1.00

### 🔸 **# Standardization**

Standardization was applied to the grade columns (G1, G2, and G3) using StandardScaler. This transformation centers the data around a mean of zero and scales it to unit variance, ensuring that all grade features contribute equally during analysis and modeling

In [21]:
# Feature Scaling using Standardization
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
df[['G1','G2','G3']] = scaler.fit_transform(df[['G1','G2','G3']])

# Verifying standardization
df[['G1','G2','G3']].describe()

G1     G2     G3
count 649.00 649.00 649.00
mean    0.00  -0.00   0.00
std     1.00   1.00   1.00
min    -4.16  -3.97  -3.69
25%    -0.51  -0.54  -0.59
50%    -0.15  -0.20   0.03
75%     0.58   0.49   0.65
max     2.77   2.55   2.20

### 🔸 **# Encoding Categorical Variables**

### 🔸 **## Label Encoding**

The categorical variable sex was converted into numerical format using Label Encoding. This transformation assigns unique numerical values to each category, making the feature suitable for statistical analysis and machine learning models

In [ ]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df['sex'] = le.fit_transform(df['sex'])

### 🔸 **# Final Clean Dataset Check**

## 🧹 **2.7 Exporting Cleaned Dataset**



The dataset has been fully preprocessed and is now ready for further analysis. The cleaned dataset is exported as a CSV file for use in the next notebook focused on Exploratory Data Analysis (EDA).



**Export Details:**



- **Format:** CSV (comma-separated values)



- **Filename:** Student_Performance_Cleaned.csv



- **Location:** "C:\Users\anton\Downloads\archive\Student_Performance_Cleaned.csv"



- **Records:** 649 students



- **Columns:** 33 features

## 🔸  Final Verdict:**



- Missing values checked and handled



- Duplicate records verified and removed (none found)



- Data types corrected where required



- Text data cleaned and standardized



- Derived features created for improved analysis



- Feature scaling applied using normalization and standardization



- Categorical variables encoded into numerical format



- Final dataset validated and confirmed ready for analysis

In [22]:
df.info()
df.isnull().sum()
df.head()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 649 entries, 0 to 648
Data columns (total 36 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   school       649 non-null    object 
 1   sex          649 non-null    object 
 2   age          649 non-null    float64
 3   address      649 non-null    object 
 4   famsize      649 non-null    object 
 5   Pstatus      649 non-null    object 
 6   Medu         649 non-null    int64  
 7   Fedu         649 non-null    int64  
 8   Mjob         649 non-null    object 
 9   Fjob         649 non-null    object 
 10  reason       649 non-null    object 
 11  guardian     649 non-null    object 
 12  traveltime   649 non-null    int64  
 13  studytime    649 non-null    int64  
 14  failures     649 non-null    int64  
 15  schoolsup    649 non-null    object 
 16  famsup       649 non-null    object 
 17  paid         649 non-null    object 
 18  activities   649 non-null    object 
 19  nursery 

age   Medu   Fedu  traveltime  studytime  failures  famrel  freetime  \
count 649.00 649.00 649.00      649.00     649.00    649.00  649.00    649.00   
mean    0.00   2.51   2.31        1.57       1.93      0.22    3.93      3.18   
std     0.04   1.13   1.10        0.75       0.83      0.59    0.96      1.05   
min     0.00   0.00   0.00        1.00       1.00      0.00    1.00      1.00   
25%     0.00   2.00   1.00        1.00       1.00      0.00    4.00      3.00   
50%     0.00   2.00   2.00        1.00       2.00      0.00    4.00      3.00   
75%     0.00   4.00   3.00        2.00       2.00      0.00    5.00      4.00   
max     1.00   4.00   4.00        4.00       4.00      3.00    5.00      5.00   

       goout   Dalc   Walc  health  absences     G1     G2     G3  avg_score  \
count 649.00 649.00 649.00  649.00    649.00 649.00 649.00 649.00     649.00   
mean    3.18   1.50   2.28    3.54      0.11   0.00  -0.00   0.00      11.63   
std     1.18   0.92   1.28    1.45      0.15   1.00   1.00   1.00       2.83   
min     1.00   1.00   1.00    1.00      0.00  -4.16  -3.97  -3.69       1.33   
25%     2.00   1.00   1.00    2.00      0.00  -0.51  -0.54  -0.59      10.00   
50%     3.00   1.00   2.00    4.00      0.06  -0.15  -0.20   0.03      11.67   
75%     4.00   2.00   3.00    5.00      0.19   0.58   0.49   0.65      13.33   
max     5.00   5.00   5.00    5.00      1.00   2.77   2.55   2.20      18.67   

       total_score  
count       649.00  
mean         34.88  
std           8.50  
min           4.00  
25%          30.00  
50%          35.00  
75%          40.00  
max          56.00

**FROM START TO DATA PREPROCESSING**:

The project successfully completed the data pre-processing stage by loading the dataset into a Pandas DataFrame and performing an initial exploration using head(), shape, info(), and describe() to understand its structure and data types. Missing values were identified and handled appropriately using statistical methods, and duplicate records were checked and removed to ensure data quality. Necessary formatting corrections were applied, including data type conversions and text standardization. Derived features such as total and average student scores were created to enhance analytical insights. Finally, numerical variables were transformed using normalization techniques to bring values to a common scale. At this stage, the dataset is clean, consistent, and ready for exploratory data analysis and visualization

**NEXT PHASE (Overview)**:

The next phase focuses on Exploratory Data Analysis (EDA), where the cleaned data will be analyzed to uncover patterns and trends in sales and profit across categories, regions, and time periods. This will involve aggregation, grouping, and visualizations such as bar charts, line graphs, and scatter plots. The final steps will summarize key insights and conclusions, translating analytical results into meaningful business interpretations and recommendations suitable for decision-making.